pkg/kubevirt/utils.go

```diff
@@ -138,26 +142,13 @@ func buildVirtualMachineInstanceTemplate(ctx *context.MachineContext) *kubevirtv
 
 	template.Spec = *ctx.KubevirtMachine.Spec.VirtualMachineTemplate.Spec.Template.Spec.DeepCopy()
 
-	cloudInitVolumeName := "cloudinitvolume"
-	cloudInitVolume := kubevirtv1.Volume{
-		Name: cloudInitVolumeName,
-		VolumeSource: kubevirtv1.VolumeSource{
-			CloudInitConfigDrive: &kubevirtv1.CloudInitConfigDriveSource{
-				UserDataSecretRef: &corev1.LocalObjectReference{
-					Name: *ctx.Machine.Spec.Bootstrap.DataSecretName + "-userdata",
-				},
-			},
-		},
-	}
-	template.Spec.Volumes = append(template.Spec.Volumes, cloudInitVolume)
-
-	cloudInitDisk := kubevirtv1.Disk{
-		Name: cloudInitVolumeName,
-		DiskDevice: kubevirtv1.DiskDevice{
-			Disk: &kubevirtv1.DiskTarget{
-				Bus: "virtio",
-			},
-		},
+	cloudInitType, index := noCloudVolume(template)
+	cloudInitVolume := cloudinitVolume(template, ctx, cloudInitType)
+	cloudInitDisk := cloudinitDisk(template, cloudInitType)
+	if cloudInitType == "CloudInitNoCloud" {
```

Comment:
instead of keying off the "CloudInitNoCloud" type here, could we instead detect if a previous cloudinit volume existed and use that index... So maybe have the `detectCloudInitVolume` function return either an index >= 0 if a volume already exists, or maybe -1 if no previous volume exists. Then with this if/def we'd just have `if index >= 0 {`

Reply:
I think while that would make it less, I believe that would require extra explaining and documentation.
It might be clearer with outputs like `exists` and `index`.

Reply:
In any case, the goal is to have it default to _CloudInitConfigDrive_, unless there's an non-nil _CloudInitNoCloud_ volume named _cloudinitvolume_

Reply:
yeah, an `exists` bool with an `index` makes sense as well.

I think i'd rather avoid detection logic that keys for the `CloudInitNoCloud` type directly, and instead have one that simply returns the cloud init type found in the volume. Right now we only have CloudInitNoCloud, and ConfigDrive, but in the future there might be more options. It would be nice if this logic naturally worked if the cloud init types are added to in the future

Reply:
Using a different method based on the type detected, such as _CloudInitNoCloud_ or _CloudInitConfigDrive_, will future proof it more than detecting only _CloudInitNoCloud_.

I'm sure there's some way to add more support another way